// @progress/kendo-angular-popup/dist/npm/services/position.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var dom_service_1 = require("./dom.service");
var util_1 = require("../util");
/**
 * @hidden
 */
var PositionService = (function () {
    function PositionService(_dom) {
        this._dom = _dom;
    }
    PositionService.prototype.positionElement = function (settings) {
        var anchor = settings.anchor, currentLocation = settings.currentLocation, positionMode = settings.positionMode, element = settings.element, anchorAlign = settings.anchorAlign, elementAlign = settings.elementAlign, collisions = settings.collisions;
        var dom = this._dom;
        var relative = dom.useRelativePosition(element);
        var viewPort = settings.viewPort || dom.windowViewPort(element);
        var anchorRect = util_1.eitherRect(dom.offset(anchor), currentLocation);
        var beforeRestrictionOffset = !relative ? currentLocation : null;
        var afterRestrictionOffset = relative ? currentLocation : null;
        var initialElementRect = util_1.replaceOffset(dom.staticOffset(element), beforeRestrictionOffset);
        var elementRect = positionMode === 'fixed' ?
            initialElementRect :
            dom.removeScroll(initialElementRect, dom.scrollPosition(element));
        var result = this._dom.restrictToView({
            anchorAlign: anchorAlign,
            anchorRect: anchorRect,
            collisions: collisions,
            elementAlign: elementAlign,
            elementRect: elementRect,
            viewPort: viewPort
        });
        return {
            flipped: result.flipped,
            offset: this._dom.addOffset(util_1.replaceOffset(initialElementRect, afterRestrictionOffset), result.offset)
        };
    };
    return PositionService;
}());
PositionService.decorators = [
    { type: core_1.Injectable },
];
/** @nocollapse */
PositionService.ctorParameters = function () { return [
    { type: dom_service_1.DOMService, },
]; };
exports.PositionService = PositionService;
